### ISSUE TITLE
Decentralised multiplayer

### ISSUE THREAD

**Issue by GrimPixel** [drwhut/tabletop-club#230]
**Is your feature request related to a problem? Please describe.**
Currently, the game's multiplayer relies on Duck DNS, which is a centralised.

**Describe the solution you'd like**
Add federated and P2P options for multiplayer.

**Comment by GrimPixel** [comment]
There is a possibility that Duck DNS goes down.
In the ideal situation, a player can simply send an invitation code that expires shortly to others, and everything including connections and calculations of physics is done on the host's computer.
Is it possible to be like this: the host can also detect disconnections, instead of relying on a master server?

**Comment by drwhut** [comment]
> There is a possibility that Duck DNS goes down.

This is true, and the master server can also go down too - but this is a risk I'm currently willing to take since a. I'm hosting this master server for free, and b. the likelihood of either service going down at any given time is slim.

> In the ideal situation, a player can simply send an invitation code that expires shortly to others, and everything including connections and calculations of physics is done on the host's computer.
> Is it possible to be like this: the host can also detect disconnections, instead of relying on a master server?

The host performs all of the physics simulation, and simply sends the results to the other clients.

In terms of the connections, unfortunately we cannot rely solely on the host to detect disconnects - the reason I keep the connection alive with the master server is because under specific circumstances, Godot by itself is not able to detect if a client has broken their connection (this usually happened under two circumstances: if too much data was sent over the wire at once, or if the client's internet connection was suddenly broken).
